PipeWire Virtual Camera

columbarius opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->
This MR creates a new plugin exposing the obs video feed as a virtual camera via PipeWire. Additionally it should create a framework to share code between different plugins connecting obs to PipeWire.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->
The current virtual camera plugin for linux via v4l2 requires root access to load a kernel driver. A plugin via PipeWire doesn't need that kind of privileges and additionally works nice in sandboxed environments like flatpak.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
* Remove the v4l2loobbak kernel module.
* Start obs
* Start the virtual camera
* Record the stream with cheese or get the pipewire-id and use the video-play example from PipeWire

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

Requires and contains: #5383 


Maintainer edit: Fixes #4499

columbarius commented:
Note:
gstreamer based clients like cheese will reconnect after a previous stream was stopped, but since the PipeWire-gstreamer plugin doesn't support restarting, the stream will be stalled until the gstreamer based node was removed.

columbarius commented:
Sry wanted to rename the branch, but looks like github doesn't support that with open MRs...

kkartaltepe commented on plugins/linux-pipewire/pipewire-virtualcam.c:
```diff
@@ -0,0 +1,318 @@
+	}
+	// copy data
+	for (uint i = 0; i < obs_pwvc->formatinfo.planes; i++) {
+		if (d[0].data == NULL) {
```
It looks like you forgot `i` here and below.

columbarius replied:
thx, fixed

dodgepong commented:
Dumb question from a non-Linux user: Are there Linux apps that expect webcams to come in via Pipewire? That is, if you're using a video chat app, for example, does it list your Pipewire video sources?

I just want to be sure we're not conflating "virtual camera" with a more generic notion of sharing OBS's primary video output with programs that want to read it. The term "Virtual Camera", as the name implies, is specifically intended for the use case of making OBS's video output appear as a system camera to applications looking for system cameras. If that isn't the use case, then we should be careful with our terminology.

tytan652 commented on plugins/linux-pipewire/pipewire-virtualcam.c:
```diff
@@ -0,0 +1,318 @@
+void virtual_cam_register_output(void)
+{
+	struct obs_output_info pipewire_virtualcam_info = {
+		.id = "virtualcam_output",
```
I think an id like "pwcam_output" or "pw_cam_output" would be better.

I don't list "pipewire_output" (that I use in #5383 as draft) because it is too confusing.
Once the name is set I change it in my PR to match.

columbarius commented:
> I see 3 solutions:
> 
>     * Use the current state with a grey stuck `start virtual cam`
> 
>     * Call `obs_output_begin_data_capture` immediately after creation and don't send them via PipeWire when state is not `_STREAMING`, this is easy to implement and what other cameras do except omitting the copy on this plugin.
> 
>     * Add a new call to obs_output like `obs_output_ready`

Should i just go for solution 2 and we may improve on it later?

pasqualirb commented:
Just for the record, https://github.com/flatpak/xdg-desktop-portal/pull/638 allow OBS restoring a session without asking for permission every time.

Conan-Kudo commented:
@GeorgesStavracas maybe you have an idea of how to resolve @columbarius' question?

Conan-Kudo commented:
Or even possibly @wtay might be able to help here.

columbarius commented:
@Conan-Kudo thanks for trying to help. 
The open question is not PipeWire related but depends on OBS internals. PipeWire has a feedback if clients are connected, which the classic webcam solutions lack. Therefore with the technically nicest solution the button will seem to be stuck until a client connects, which isn't convenient for the user unaware of this problem. Otherwise this branch should work.

Additionally there are some MRs which I like to priorities first and then work on this will continue. 

If you want to help, please run this branch and report any issue you find.

columbarius commented:
For usage with v4l2 apps:

* `sudo modprobe v4l2loopback exclusive_caps=1 video_nr=X card_label="OBS PipeWire Virtual Cam"`
* start obs from terminal and look in the log for the PipeWire stream id after the Virtual Camera was started.
* `gst-launch-1.0 pipewiresrc path=pw_id ! videoconvert ! v4l2sink device=/dev/videoX`

Replaying this with
`gst-launch-1.0 v4l2src device=/dev/videoX ! videoconvert ! xvimagesink`
works fine but other applications like firefox seemed to have problems.

Conan-Kudo commented:
@columbarius that's not required. PipeWire has a `pw-v4l2` program you can use to make PipeWire camera objects show up in apps as v4l2 devices.

`pw-v4l2 <program>` is enough for that.

columbarius commented:
@Conan-Kudo ahh looks like it now works with firefox, but with vlc I'm still out of luck.

columbarius commented:
rebased onto #5394 which should be merged soon.

GeorgesStavracas commented:
Sorry, looks like this will need another rebase!

tytan652 commented:
I made a mistake and completely forgot to test my dual vcam PR when one of them doesn't start for some reason (eg. pkexec canceled).
So I need to work on it to fix it.

columbarius commented:
rebased onto #5394 which should be merged soon.

> I made a mistake and completely forgot to test my dual vcam PR when one of them doesn't start for some reason (eg. pkexec canceled).

I've removed v4l2-loopback from my system and this is working for me.

tytan652 commented:
#5383 is now out of draft, I fixed all the issues that I found. Please rebase on to fix them here too.

columbarius commented:
May this receive the `seeking testers` label?

columbarius commented:
Thanks!

tytan652 commented:
Note: On "Use Both" the button will not get stuck waiting for a PipeWire connection because it's V4L2 VCam that change the button state.

But on PipeWire only, the button will be stuck waiting for a connection.

columbarius commented:
Anything else blocking this?

tytan652 commented:
> Anything else blocking this?

My PR did not get any reviews, so until mine is not merged, this one is stuck.

Normally your PR should be drafted waiting for my PR to be merged first.

Fenrirthviti commented:
Marked as draft until the dependent change can be reviewed.

columbarius commented:
> > Anything else blocking this?
> 
> My PR did not get any reviews, so until mine is not merged, this one is stuck.

Should your PR include a possible solution for the "stuck" start virtualcam button, or should this be added later?

tytan652 commented:
> Should your PR include a possible solution for the "stuck" start virtualcam button, or should this be added later?

The "stuck button" can not be fixed on the UI side since the button should only get "unstuck" by a signal emitted by the output. To fix this something should be done on libobs side.

tytan652 commented on plugins/linux-pipewire/pipewire.c:
```diff
@@ -396,42 +462,87 @@ static void init_format_info(obs_pipewire_data *obs_pw)
+
+	uint32_t output_flags;
+
+	output_flags = obs_source_get_output_flags(obs_pw_stream->source);
```
I had to do this to make it rework.
```suggestion
	output_flags = obs_source_get_output_flags(obs_pw_stream->obs_data.source);
```

tytan652 commented on plugins/linux-pipewire/pipewire-virtualcam.c:
```diff
@@ -0,0 +1,124 @@
+	stream_data.type = OBS_PIPEWIRE_STREAM_TYPE_OUTPUT;
+	stream_data.output = obs_pwvc->output;
+	obs_pwvc->obs_pw_stream = obs_pipewire_connect_stream(
+		obs_pwvc->obs_pw, &stream_data, PW_ID_ANY, "OBS Virtual Camera",
+		pw_properties_new(PW_KEY_NODE_DESCRIPTION, "OBS Virtual Camera",
```
I did the same to V4L2, to be able to identify them.
```suggestion
		obs_pwvc->obs_pw, &stream_data, PW_ID_ANY, "OBS Virtual Camera (PipeWire)",
		pw_properties_new(PW_KEY_NODE_DESCRIPTION, "OBS Virtual Camera (PipeWire)",
```

tytan652 reviewed (commented):
I reworked #5383, so it needs a rebase. And here some nitpik.

columbarius commented:
> I reworked #5383, so it needs a rebase. And here some nitpik.

Thanks. I reworked the whole #6669 quite extreme, so I'd like to get some feedback on it before rebasing this again. At the end of the MR my dev branch is linked. Feel free to take a look at it please!

columbarius committed:
linux-pipewire: Extract helper to manage supported formats

columbarius committed:
linux-pipewire: Add pipewire based virtualcamera

columbarius committed:
DNM: linux-pipewire: Extract internal symbols from pipewire.c

columbarius committed:
linux-pipewire: Allow connecting to pipewire without fd

columbarius committed:
DNM: linux-{pipewire,v4l2}: replace v4l2 with pw virtualcam

columbarius commented:
Since this MR will be blocked for a while (waiting for extensions to the portal spec). I restructured it extracting a shared helper and made it selfcontained so rebasing should be much easier. The ui parts to select the output are dropped in favor of two DNM commits, which contain hacks getting this running and testable.

When all dependencies are done those should be replaced with the correct integration.

columbarius commented:
Note: I chose 640x480 as the preferred resolution of the stream. 1920x1080 worked, but lagged a little bit (shm buffers with memcpy are problaby not that efficient).

PatTheMav commented:
As mentioned in https://github.com/obsproject/obs-studio/pull/6669, can someone give me a conceptual (non-technical) breakdown on what the issue with this PR is, mainly why it's in a Draft state (suggesting it's not ready to be reviewed) and why it's been in limbo for almost 4 years now?

raploz commented:
> > Since this MR will be blocked for a while (waiting for extensions to the portal spec).
> 
> Are there any updates on this? Is there an MR or discussion in the xdg desktop portal repo?

I think the discussion about the architecture for a portal is happening here: https://github.com/flatpak/xdg-desktop-portal/discussions/1141

> As mentioned in #6669, can someone give me a conceptual (non-technical) breakdown on what the issue with this PR is, mainly why it's in a Draft state (suggesting it's not ready to be reviewed) and why it's been in limbo for almost 4 years now?

The portal seems to be necessary to make everything work, but I don't know the technical details. The link above should have some details.

columbarius commented:
Iirc there where two issues:
* No option to select the virtual camera plugin (v4l2 or pipewire)
* Without a portal, this requires mounting the pipewire socket as a flatpak and the idea was to find some portal solution for app2app videosharing first.
There might be changes since i wasn`t able to follow the discussion anymore. 
If the first point is solved (or just prefering this over the v4l2 solution whenever a pipewire socket is available) and the second is solved or not a requirement anymore, I cam try to find some time to update this.

parametalol commented:
> add an implementation which would depend on unsandboxed pipewire access.

Isn't it fine to depend on unsandboxed pipewire access if there is no alternative for such an important feature until proper solution is available? Just let the user know the trade-offs.

jp-bennett commented:
This is a feature that many users are desperately looking for, and the reason for holding it up rings quite hollow when the alternative is compiling an out-of-tree kernel module.
